### DISCUSSION TITLE
MQTT Topic Structure in MTConnect .NET Agent Version 6.0.5 and above.

### DISCUSSION THREAD

**Discussion by Pranay265** [TrakHound/MTConnect.NET#58] [Q&A] [answered]
I am currently working with the MTConnect .NET Agent version 6.0.5 and have encountered an issue related to the MQTT topic structure. I am attempting to replicate the topic structure that was achieved in version 5.4.0, specifically the structure **MTConnect/Devices/{deviceuuid}/Observations/{dataitemId}**. However, despite making adjustments to the configuration file, I am still receiving MQTT topics such as probe, sample, and current instead of the desired observations structure.

Comparison to Version 5.4.0:

In version 5.4.0 of the MTConnect .NET Agent, the configuration and MQTT topic structure were set up as follows:

#  MQTT Configuration -

server: localhost
port: 1883
mqttFormat: Flat
This configuration produced MQTT topics in the format MTConnect/Devices/{deviceuuid}/Observations/{dataitemId} as expected.

However, in version 6.0.5 of the MTConnect .NET Agent, I have made the following adjustments to the configuration file to replicate the same topic structure:

# - MQTT Configuration -

- mqtt-relay:
    server: localhost
    port: 1883
    topicPrefix: MTConnect/Devices/{deviceuuid}
    mqttFormat: Flat
   
Despite these adjustments, I am still receiving MQTT topics such as probe, sample, and current instead of the desired observations structure.

I kindly request assistance in identifying the correct configuration settings or any additional steps required to achieve the MQTT topic structure MTConnect/Devices/{deviceuuid}/Observations/{dataitemId} in MTConnect .NET Agent version 6.0.5. Any insights or guidance on resolving this issue would be greatly appreciated. Thank you!

**Comment by PatrickRitchie** [comment, marked as answer]
I just published a new release [v6.2.1](https://github.com/TrakHound/MTConnect.NET/releases/tag/v6.2.1-beta) that adds the ability to toggle between the "Document" and "Entity" topic structure using the below configuration:

```yaml
- mqtt-relay:
    server: localhost
    port: 1883
    topicPrefix: MTConnect
    topicStructure: Entity
    documentFormat: json
```

Let me know if that works for you or if you have any other questions.

**Reply by Pranay265** [reply]
Thanks a lot, @PatrickRitchie , for your prompt response. The above solution provided worked for me, and the MQTT topic structure now matches exactly as requested. However, upon further inspection, I noticed a slight variation in the payload format compared to the previous version.

**Version 6.2.1 Payload:**
{
  "dataItemId": "string",
  "name": "string",
  "type": "string",
  "timestamp": "2024-04-03T10:46:04.003106Z",
  "result": "UNAVAILABLE"
}

**Version 5.4.0 Payload:**
{
  "dataItemId": "string",
  "name": "string",
  "**category**": "string",
  "type": "string",
  "timestamp": "2024-03-22T11:34:37.8193737Z",
  "**sequence**": "59," -- version 6.1.3
  "**instanceId**": 1711107277,
  "result": "UNAVAILABLE"
}
I'm curious if there are any changes or if some data items are skipped in the latest version that might explain this variation in the payload structure. Your insights would be greatly appreciated as I strive to ensure consistency and accuracy in the data received. Thank you for your assistance!

**Reply by PatrickRitchie** [reply]
Glad to hear that fixed the initial issue. I just uploaded another release [v6.2.2](https://github.com/TrakHound/MTConnect.NET/releases/tag/v6.2.2-beta) that should add the sequence, category, and instanceId properties. I also made condition observations a JSON array, as there may be multiple condition fault states active at a single time. I may separate the dataitems into seperate Sample, Event, and Condition topics to help when reading from a client.

Let me know if you find anything else or have any other questions.

**Reply by Pranay265** [reply]
Thank you very much for addressing the issue promptly and for providing insights into future changes. I'm pleased to confirm that the problem has been resolved. As of now, everything looks good, and I will definitely reach out if any further queries or issues arise. Your support is greatly appreciated!
